libraries/chain/include/eosio/chain/transaction_context.hpp

```diff
@@ -37,6 +37,16 @@ namespace eosio { namespace chain {
        uint64_t used_cpu;
    };
 
+   struct ram_providers {
+
+      void add_provider(account_name contract, account_name user, account_name provider);
+
+   private:
+       using avalable_providers_for_user = std::map<account_name, std::vector<account_name>>;
+
+       std::map<account_name, avalable_providers_for_user>  available_providers_;
```

Comment:
It's seems that available_providers_ more complex then actually need.
Maybe better `std::map<std::pair<contract,account>, provider>`?
What behavior is planned if for one contract, account there will be several providers? I thin this is an erroneous transaction.